File: c++/84. Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        vector<int> &a = heights;
        if (a.empty()) return 0;
        int n = a.size();
        vector<int> ll(n, 0), rr(n, 0); // ll, rr 分别记录当前位置i能够向左和向右延伸的最大距离
        
        int i , j;
        ll[0] = 0; 
        for (i = 1; i < n; i++) {
            ll[i] = i;
            j = i;
            while (ll[j] - 1 >= 0 && a[ll[j]-1] >= a[i]) // if左边的高度大于等于a[i]，向左延伸一步
                j = ll[j] - 1;
            ll[i] = ll[j];
        }
        
        rr[n-1] = n-1;
        for (i = n-2; i >= 0; i--) {
            rr[i] = i;
            j = i;
            while (rr[j] + 1 < n && a[rr[j] + 1] >= a[i])
                j = rr[j] + 1;
            rr[i] = rr[j];
        }
        
        int res = 0;
        for (int i = 0; i < n; i++) {
            res = max(res, (rr[i] - ll[i] + 1) * a[i]);
        }
        
        return res;
    }
};

File: python/84. Largest Rectangle in Histogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        res = 0
        heights = [0] + heights + [0]
        stack = []
        for i, h in enumerate(heights):
            while stack and heights[stack[-1]] > h:
                top = stack.pop()
                res = max(res, heights[top]*(i-stack[-1]-1))
            stack.append(i)
        

        return res

# 思路：维持一个单调递增栈   
# 这道题考虑用一个栈来维持一个递增序列，栈中保存的是当前元素的索引。如果当前值比栈顶元素小，那么栈中元素就需要弹出，直到栈为空或者栈顶元素小于当前元素。这样的话可以保证栈中第i个元素对应的高度总是小于等于栈中第i-1个元素至当前欲插入元素对应的矩阵高度，这样当元素弹出时，以它为最低高度的最大矩阵面积就很好求出来了。高直接可以得出，而长度则为第i-1个元素对应的索引至当前欲插入元素的索引的差值。
# ————————————————
# 版权声明：本文为CSDN博主「沧海漂游_」的原创文章，遵循 CC 4.0 BY-SA 版权协议，转载请附上原文出处链接及本声明。
# 原文链接：https://blog.csdn.net/lv1224/article/details/79974175

        
if __name__ == '__main__':
    heights = [0, 9]
    su = Solution()
    print su.largestRectangleArea(heights)